### PR TITLE
[HW] HWModule: store input port locations only on block args

### LINKED ISSUE #5484: circt-opt cannot consume the generic format it generates
`circt-opt` cannot take as input the output it emitted using `-mlir-print-op-generic`. 

The reason this happens seems to be because of the location attributes attached to HwModule. Are they truly necessary? It seems to fail because they are inconsistent with block arguments. But is this check necessary? And if the locations can be found from the block arguments, are the explicit locations necessary?

## Instructions to reproduce

Give the following snippet to `circt-opt -mlir-print-op-generic`:

```mlir
hw.module @B(%a: i1) -> (nameOfPortInSV: i1, "": i1) {
  %0 = comb.or %a, %a : i1
  %1 = comb.and %a, %a : i1
  hw.output %0, %1: i1, i1
}
```

This should yield something like:
```
#loc = loc("<stdin>":1:14)
#loc1 = loc("<stdin>":1:26)
#loc2 = loc("<stdin>":1:46)
"builtin.module"() ({
  "hw.module"() ({
  ^bb0(%arg0: i1):
    %0 = "comb.or"(%arg0, %arg0) : (i1, i1) -> i1
    %1 = "comb.and"(%arg0, %arg0) : (i1, i1) -> i1
    "hw.output"(%0, %1) : (i1, i1) -> ()
  }) {argLocs = [#loc], argNames = ["a"], comment = "", function_type = (i1) -> (i1, i1), parameters = [], resultLocs = [#loc1, #loc2], resultNames = ["nameOfPortInSV", ""], sym_name = "B"} : () -> ()
}) : () -> ()
```

Giving this output to `circt-opt` again yields an error:
```
<stdin>:5:3: error: 'hw.module' op block argument locations should match signature locations
  "hw.module"() ({
  ^
```

### PR DESCRIPTION
**Pull request by youngar**
On modules we keep an array attribute `port_locs` for port location information.  On regular HWModules specifically, which have a body block, we copy the input port location to the entry block arguments, for error messages and such. This is violating the "single source of truth" principle for input port location info, so we added a verifier that ensured the block argument locations were always kept in sync with the port location attributes. This lead to another problem: if you print our IR in the generalized format and try to read it back in, by default with `-mlir-print-debug-info` printing turned off, it will end up throwing away the location on the blocks. When the IR is read back in, the block arguments will have new location information corresponding to the intermediate file, while the location array maintains the old locations. This trips the verifier since the port attribute location will be out of sync with the block argument location.

To fix this problem, this changes the HWModule op to record input port locations only on the blocks, and to store the output port locations in a new `result_locs` array.  The other module types remain the same.  Now, when printing in the generic format, input ports will lose their original location information, but the result locations will remain the same.  If printing with debug information, then both will retain their original locations.

As an example:
```mlir
// locs.mlir
hw.module @B(in %a: i1, out nameOfPortInSV: i1) {
  hw.output %a: i1
}
```
Printing without the debug information with`circt-opt -mlir-print-op-generic locs.mlir | ./bin/circt-opt --mlir-print-debuginfo` gives:
```mlir
#loc2 = loc("<stdin>":4:8)
#loc3 = loc("locs.mlir":1:29)
module {
  hw.module @B(in %a : i1 loc(#loc2), out nameOfPortInSV : i1 loc(#loc3)) {
    hw.output %a : i1 loc(#loc4)
  } loc(#loc1)
} loc(#loc)
#loc = loc("<stdin>":2:1)
#loc1 = loc("<stdin>":3:3)
#loc4 = loc("<stdin>":5:5)
```
Printing with the debug information with `circt-opt -mlir-print-op-generic --mlir-print-debuginfo locs.mlir | ./bin/circt-opt --mlir-print-debuginfo` gives:
```mlir
#loc2 = loc("locs.mlir":1:17)
#loc3 = loc("locs.mlir":1:29)
module {
  hw.module @B(in %a : i1 loc(#loc2), out nameOfPortInSV : i1 loc(#loc3)) {
    hw.output %a : i1 loc(#loc4)
  } loc(#loc1)
} loc(#loc)
#loc = loc("locs.mlir":0:0)
#loc1 = loc("locs.mlir":1:1)
#loc4 = loc("locs.mlir":2:3)
```

Closes #5484.

### CODE REVIEW

**Review by darthscsi — COMMENTED** [review]
I was trying hard to get completely away from having any properties which treated input and output ports differently.  This causes (as you had to write) lots of extra code.  It would be nice if we could go the other way and not store locations on blockargs.  I don't see a great way to do that though.

I have thought we should have output ports be inout<> block args and eliminate hw.output.  That would also achieve these goals.

**Comment by teqdruid** [comment]
> I was trying hard to get completely away from having any properties which treated input and output ports differently. This causes (as you had to write) lots of extra code. It would be nice if we could go the other way and not store locations on blockargs. I don't see a great way to do that though.

I agree with the goal, but I don't see a great way either. The problem is really just the verifier, correct? We could just special case it to check if the port_locs attr is empty and just not throw an error. That may be treating the symptom rather than the cause, however.

> I have thought we should have output ports be inout<> block args and eliminate hw.output. That would also achieve these goals.

Something about that makes my skin crawl. I really like how modules look a lot like functions, but now that we have ModuleLike instead of FunctionLike, the similarity is just superficial I think. (If we go this way, I would strongly prefer a new `output<>` type and have it be compatible with the existing hw.instance call-like/dataflow format.)

What benefits would this have other than unifying port_locs and block arg locs?

**Comment by youngar** [comment]
We talked about this PR in [today's ODM meeting](https://docs.google.com/document/d/1fOSRdyZR2w75D87yU2Ma9h2-_lEPL4NxvhJGJd-s5pk/edit#heading=h.9x3hw8z1qyo9), the plan is to go ahead with this PR.  I will get some more review on this and then get it merged.

File: lib/Dialect/HW/HWOps.cpp [hunk 1]
```diff
@@ -747,6 +740,13 @@ void HWModuleOp::build(OpBuilder &builder, OperationState &result,
     body->addArgument(type, loc);
   }
 
+  // Add result ports attribute.
```
Comment (darthscsi) [inline lib/Dialect/HW/HWOps.cpp:743]: This adds 2 more iterations over the ports.  Can't this be done in the previous loop?

File: lib/Dialect/HW/HWOps.cpp [hunk 2]
```diff
@@ -872,40 +886,12 @@ static bool hasAttribute(StringRef name, ArrayRef<NamedAttribute> attrs) {
...
-  if (llvm::any_of(attrs, nonEmptyAttrsFn))
-    result.addAttribute(portAttrsName, builder.getArrayAttr(attrs));
-
-  if (llvm::any_of(locs, nonEmptyLocsFn))
-    result.addAttribute(portLocsName, builder.getArrayAttr(locs));
-}
-
 template <typename ModuleTy>
-static ParseResult parseHWModuleOp(OpAsmParser &parser, OperationState &result,
-                                   ExternModKind modKind = PlainMod) {
```
Comment (darthscsi) [inline lib/Dialect/HW/HWOps.cpp:906]: Can you factor out the ExternModKind change to a separate commit?

**Pull request merged**